README.md
## dc.parallel-coordinates.js

Experiment of integrating [syntagmatic/parallel-coordinates](https://github.com/syntagmatic/parallel-coordinates) (via [BigFatDog/parcoords-es](https://github.com/BigFatDog/parcoords-es)) with [dc.js](https://github.com/syntagmatic/parallel-coordinates).

This doesn't work yet - all the functionality is there but the quirks of the two libraries need to be aligned.


Changelog.md
## dc.parallel-coordinates.js 0.0.1

Nothing here yet.



web/cars.js
d3.csv('cars.csv').then(function(data) {
    var numeric = ['cylinders','displacement (cc)','power (hp)','weight (lb)','0-60 mph (s)','year'];
    data.forEach(function(d) {
        numeric.forEach(function(f) {
            d[f] = +d[f];
        });
    });

    var cf = crossfilter(data);
    var car_key = d => d.name + ' ' + d.year;
    var nameDim = cf.dimension(car_key);

    var par = dc_parcoords.parallelCoords('#parcoords');
    var blue_to_brown = d3.scaleLinear()
            .domain([9, 50])
            .range(["steelblue", "brown"])
            .interpolate(d3.interpolateLab);
    par
        .width(600)
        .height(300)
        .dimension(nameDim)
        .keyAccessor(car_key)
        .colorAccessor(d => d['economy (mpg)'])
        .colors(blue_to_brown)
        .hideAxis(['name']);

    par.parcoords().margin({top:20, left:50, right:20, bottom: 20});

    var yearDim = cf.dimension(d => d.year),
        yearWeightGroup = yearDim.group().reduceSum(d => d['weight (lb)']);

    var pie = dc.pieChart('#pie');
    pie
        .width(250)
        .height(250)
        .dimension(yearDim)
        .group(yearWeightGroup)
        .minAngleForLabel(0.01);

    var powerDim = cf.dimension(d => d['power (hp)']),
        powerAccelGroup = powerDim.group(p => Math.floor(p/25)*25).reduce(
            function(p, v) { // add
                ++p.n;
                p.tot += v['0-60 mph (s)'];
                return p;
            },
            function(p, v) { // remove
                --p.n;
                p.tot -= v['0-60 mph (s)'];
                return p;
            },
            function() {
                return {n: 0, tot: 0};
            }
        );
    var line = dc.lineChart('#line');
    line
        .width(400)
        .height(200)
        .xAxisLabel('power (hp)')
        .yAxisLabel('avg 0-60 mph (s)')
        .dimension(powerDim)
        .group(powerAccelGroup)
        .valueAccessor(kv => kv.value.n && kv.value.tot/kv.value.n)
        .x(d3.scaleLinear()).elasticX(true)
        .elasticY(true);

    dc.renderAll();

    par.parcoords()
        .brushMode('1D-axes');
});


Gruntfile.js
module.exports = function (grunt) {
    'use strict';

    require('load-grunt-tasks')(grunt, {
        pattern: ['grunt-*', '!grunt-lib-phantomjs']
    });
    require('time-grunt')(grunt);

    var config = {
        src: 'src',
        spec: 'spec',
        web: 'web',
        pkg: require('./package.json'),
        banner: grunt.file.read('./LICENSE_BANNER'),
        jsFiles: module.exports.jsFiles
    };

    grunt.initConfig({
        conf: config,

        concat: {
            options : {
                process: true,
                sourceMap: true,
                banner : '<%= conf.banner %>'
            },
            js: {
                src: '<%= conf.jsFiles %>',
                dest: '<%= conf.pkg.name %>.js'
            }
        },
        uglify: {
            jsmin: {
                options: {
                    mangle: true,
                    compress: true,
                    sourceMap: true,
                    banner : '<%= conf.banner %>'
                },
                src: '<%= conf.pkg.name %>.js',
                dest: '<%= conf.pkg.name %>.min.js'
            }
        },
        jscs: {
            old: {
                src: ['<%= conf.spec %>/**/*.js'],
                options: {
                    validateIndentation: 4
                }
            },
            source: {
                src: ['<%= conf.src %>/**/*.js', '!<%= conf.src %>/{banner,footer}.js', 'Gruntfile.js',
                      'grunt/*.js', '<%= conf.web %>/stock.js'],
                options: {
                    config: '.jscsrc'
                }
            }
        },
        jshint: {
            source: {
                src: ['<%= conf.src %>/**/*.js', 'Gruntfile.js', 'grunt/*.js', '<%= conf.web %>/stock.js'],
                options: {
                    jshintrc: '.jshintrc',
                    ignores: ['<%= conf.src %>/banner.js', '<%= conf.src %>/footer.js']
                }
            }
        },
        watch: {
            scripts: {
                files: ['<%= conf.src %>/**/*.js'],
                tasks: ['docs']
            },
            tests: {
                files: ['<%= conf.src %>/**/*.js', '<%= conf.spec %>/**/*.js'],
                tasks: ['test']
            },
            reload: {
                files: ['<%= conf.pkg.name %>.js',
                        '<%= conf.pkg.name %>css',
                        '<%= conf.web %>/js/<%= conf.pkg.name %>.js',
                        '<%= conf.web %>/css/<%= conf.pkg.name %>.css',
                        '<%= conf.pkg.name %>.min.js'],
                options: {
                    livereload: true
                }
            }
        },
        connect: {
            server: {
                options: {
                    port: 8888,
                    base: '.'
                }
            }
        },
        emu: {
            api: {
                src: '<%= conf.pkg.name %>.js',
                dest: '<%= conf.web %>/docs/api-latest.md'
            }
        },
        toc: {
            api: {
                src: '<%= emu.api.dest %>',
                dest: '<%= emu.api.dest %>'
            }
        },
        markdown: {
            html: {
                src: '<%= emu.api.dest %>',
                dest: '<%= conf.web %>/docs/index.html'
            },
            options: {markdownOptions: {highlight: 'manual'}}
        },
        copy: {
            'dc-to-gh': {
                files: [
                    {
                        expand: true,
                        flatten: true,
                        nonull: true,
                        src: [
                            '<%= conf.pkg.name %>.css',
                            'node_modules/dc/dc.css',
                            'node_modules/parcoord-es/dist/parcoords.css'
                        ],
                        dest: '<%= conf.web %>/css/'
                    },
                    {
                        expand: true,
                        flatten: true,
                        nonull: true,
                        src: [
                            '<%= conf.pkg.name %>.js',
                            '<%= conf.pkg.name %>.js.map',
                            '<%= conf.pkg.name %>.min.js',
                            '<%= conf.pkg.name %>.min.js.map',
                            'node_modules/jquery/dist/jquery.js',
                            'node_modules/d3/dist/d3.js',
                            'node_modules/dc/dc.js',
                            'node_modules/parcoord-es/dist/parcoords.standalone.js',
                            'node_modules/crossfilter2/crossfilter.js'
                        ],
                        dest: '<%= conf.web %>/js/'
                    }
                ]
            }
        },
        'gh-pages': {
            options: {
                base: '<%= conf.web %>',
                message: 'Synced from from master branch.'
            },
            src: ['**']
        },
        shell: {
            merge: {
                command: function (pr) {
                    return [
                        'git fetch origin',
                        'git checkout master',
                        'git reset --hard origin/master',
                        'git fetch origin',
                        'git merge --no-ff origin/pr/' + pr + ' -m \'Merge pull request #' + pr + '\''
                    ].join('&&');
                },
                options: {
                    stdout: true,
                    failOnError: true
                }
            },
            amend: {
                command: 'git commit -a --amend --no-edit',
                options: {
                    stdout: true,
                    failOnError: true
                }
            },
            hooks: {
                command: 'cp -n scripts/pre-commit.sh .git/hooks/pre-commit' +
                    ' || echo \'Cowardly refusing to overwrite your existing git pre-commit hook.\''
            }
        },
        browserify: {
            dev: {
                src: '<%= conf.pkg.name %>.js',
                dest: 'bundle.js',
                options: {
                    browserifyOptions: {
                        standalone: 'dc'
                    }
                }
            }
        }
    });

    // custom tasks
    grunt.registerMultiTask('emu', 'Documentation extraction by emu.', function () {
        var emu = require('emu'),
            srcFile = this.files[0].src[0],
            destFile = this.files[0].dest,
            source = grunt.file.read(srcFile);
        grunt.file.write(destFile, emu.getComments(source));
        grunt.log.writeln('File \'' + destFile + '\' created.');
    });
    grunt.registerTask('merge', 'Merge a github pull request.', function (pr) {
        grunt.log.writeln('Merge Github Pull Request #' + pr);
        grunt.task.run(['shell:merge:' + pr, 'test' , 'shell:amend']);
    });
    grunt.registerMultiTask('toc', 'Generate a markdown table of contents.', function () {
        var marked = require('marked'),
            slugify = function (s) { return s.trim().replace(/[-_\s]+/g, '-').toLowerCase(); },
            srcFile = this.files[0].src[0],
            destFile = this.files[0].dest,
            source = grunt.file.read(srcFile),
            tokens = marked.lexer(source),
            toc = tokens.filter(function (item) {
                return item.type === 'heading' && item.depth === 2;
            }).reduce(function (toc, item) {
                return toc + '  * [' + item.text + '](#' + slugify(item.text) + ')\n';
            }, '');

        grunt.file.write(destFile, '# dc.parallel-coordinates.js API\n' + toc + '\n' + source);
        grunt.log.writeln('Added TOC to \'' + destFile + '\'.');
    });
    grunt.registerTask('test-stock-example', 'Test a new rendering of the stock example web page against a ' +
                       'baseline rendering', function (option) {
                           require('./regression/stock-regression-test.js').testStockExample(this.async(), option === 'diff');
                       });
    grunt.registerTask('update-stock-example', 'Update the baseline stock example web page.', function () {
        require('./regression/stock-regression-test.js').updateStockExample(this.async());
    });

    // task aliases
    grunt.registerTask('build', ['concat', 'uglify']);
    grunt.registerTask('docs', ['build', 'copy', 'emu', 'toc', 'markdown']);
    grunt.registerTask('web', ['docs', 'gh-pages']);
    grunt.registerTask('server', ['docs', 'connect:server', 'watch:scripts']);
    grunt.registerTask('test', ['build', 'jasmine:specs', 'shell:hooks']);
    grunt.registerTask('test-browserify', ['build', 'browserify', 'jasmine:browserify']);
    grunt.registerTask('coverage', ['build', 'jasmine:coverage']);
    grunt.registerTask('ci', ['test', 'jasmine:specs:build', 'connect:server', 'saucelabs-jasmine']);
    grunt.registerTask('ci-pull', ['test', 'jasmine:specs:build', 'connect:server']);
    grunt.registerTask('lint', ['build', 'jshint', 'jscs']);
    grunt.registerTask('default', ['build']);
};

module.exports.jsFiles = [
    'src/banner.js',   // NOTE: keep this first
    'src/core.js', // then this!
    'src/parallel-coordinates.js',
    'src/footer.js'  // NOTE: keep this last
];


src/parallel-coordinates.js
dc_parcoords.parallelCoords = function(selector, chartGroup) {
    var _chart = {}, // this object
        _parcoords, // syntagmatic parallel-coordinates object
        _root, // parent div
        _dimension; // array of crossfilter dimensions

    var _width = 400, _height = 300,
        _keyAccessor,
        _colorAccessor, _colorScale,
        _hideAxis;

    var NO;

    _chart.parcoords = function() {
        if(!_parcoords) {
            _parcoords = ParCoords()(selector);
            _parcoords.on('brush', function(d) {
                if(_dimension && _keyAccessor) {
                    if(!d || !d.length)
                        _dimension.filter(null);
                    else {
                        var keys = d.map(_keyAccessor);
                        _dimension.filter(function(k2) {
                            return keys.indexOf(k2) !== -1;
                        });
                        NO=true;
                        dc.redrawAll(chartGroup);
                        NO=false;
                    }
                }
                console.log(d);
            });
        }
        if(_dimension) {
            _parcoords
                .data(_dimension.top(Infinity));
        }
        return _parcoords;
    };

    _chart.render = function() {
        _root = d3.select(selector);
        _root
            .style('height', _height + 'px')
            .style('width', _width + 'px')
            .classed('parcoords', true);
        var parcoords = _chart.parcoords();
        parcoords
            .width(_width)
            .height(_height);
        if(_chart.hideAxis())
            parcoords.hideAxis(_chart.hideAxis());
        if(_colorAccessor) {
            if(_colorScale)
                parcoords.color(function(d) {
                    return _colorScale(_colorAccessor(d));
                });
            else
                parcoords.color(function(d) {
                    return _colorAccessor(d);
                });
        }
        return _chart.redraw();
    };

    var foo;
    _chart.redraw = function() {
        if(NO) return this;
        _parcoords
            .data(_dimension.top(Infinity))
            .render();
        if(!foo)
            _parcoords.createAxes();
        else
            _parcoords.updateAxes();
        foo = true;
        return this;
    };

    _chart.dimension = function(_) {
        if(!arguments.length)
            return _dimension;
        _dimension = _;
        return this;
    };

    _chart.width = function(_) {
        if(!arguments.length)
            return _width;
        _width = _;
        return this;
    };

    _chart.height = function(_) {
        if(!arguments.length)
            return _height;
        _height = _;
        return this;
    };

    _chart.keyAccessor = function(_) {
        if(!arguments.length)
            return _keyAccessor;
        _keyAccessor = _;
        return this;
    };

    _chart.colorAccessor = function(_) {
        if(!arguments.length)
            return _colorAccessor;
        _colorAccessor = _;
        return this;
    };

    _chart.colors = function(_) {
        if(!arguments.length)
            return _colorScale;
        _colorScale = _;
        return this;
    };

    _chart.hideAxis = function(_) {
        if(!arguments.length)
            return _hideAxis;
        _hideAxis = _;
        return this;
    };

    dc.registerChart(_chart, chartGroup);

    return _chart;
};
